PHP migration guide (#1495)
chainguard-dev/edu: content/chainguard/migration-guides/migrating-to-php-images.md

---
title: "Migrating to PHP Chainguard Images"
linktitle: "Migrating PHP"
type: "article"
description: "Guidance on how to migrate PHP Dockerfile workloads to use Chainguard Images"
date: 2024-04-04T15:56:52-07:00
lastmod: 2024-04-04T15:56:52-07:00
draft: false
tags: ["Images", "Product", "Conceptual"]
images: []
weight: 100
toc: true
---

Chainguard Images are built on top of [Wolfi](/chainguard/open-source/wolfi/), a Linux _undistro_ designed specifically for containers. Our PHP images have a minimal design that ensures a smaller attack surface, which results in smaller images [with low to zero](/chainguard/chainguard-images/vuln-comparison/php/) CVEs. Nightly builds deliver fresh images whenever updated packages are available, which also helps to reduce the toil of manually patching CVEs in PHP images.

> 

Reviewer: One general note: there isn't a lot of consistency in whether it's stylized as "Image" or "image". I think for any instance where you're talking about one or more Chainguard images, you should capitalize the "I", and in cases where you're talking about images generally, it's fine to just leave it lowercase.

Author: Do you mean for instance "Our PHP Images" vs "Our PHP images"  ? The way I have been using it so far is whenever it's together "Chainguard Images" I make sure is uppercase, because it's the product name. But if I'm referring to the unit of image (one or more, seen as objects) then I am using the lowercase.

This article will assist you in the process of migrating your existing PHP Dockerfiles to leverage the benefits of Chainguard Images, for a smaller attack surface and a more secure application footprint.

## Chainguard PHP Images

Chainguard PHP Images offer multiple variants catering to distinct use cases, such as command-line interfaces (CLI) and web applications. Each variant comes in two flavors: a minimal runtime image (distroless) and a builder image distinguished by the `-dev` suffix (e.g., `latest-dev`).

> In a nutshell, distroless images don't include a package manager or a shell, being used exclusively as runtimes to keep the environment to a minimum. Builder images, on the other hand, include packages such as `apk` and `composer` for building PHP applications. Builder images can be used _as-is_ to provide a more straightforward migration path, but we encourage users to combine both images in a multi stage environment to build a final distroless image that serves as a strict application runtime, whenever that is possible.

Author: yeah that one I am always unsure! will stick to the docker convention then

The following `Dockerfile` uses the `php:latest-dev` image to build the application, which in this case means copying the application files and installing dependencies via `composer`. A second build stage copies the application to a final distroless image based on `php:latest`.

```Dockerfile
FROM cgr.dev/chainguard/php:latest-dev AS builder
USER root
COPY . /app
RUN chown -R php /app
USER php
RUN cd /app && \
    composer install --no-progress --no-dev --prefer-dist

FROM cgr.dev/chainguard/php:latest
COPY --from=builder /app /app

ENTRYPOINT [ "php", "/app/myscript.php" ]
```

Our [PHP Getting Started](/chainguard/chainguard-images/getting-started/php/) guide has step-by-step instructions on how to build and run a PHP CLI application with Chainguard Images.

## Migrating PHP Web applications to use Chainguard Images
For PHP web applications that serve content through a web server, you should use the `latest-fpm` and `latest-fpm-dev` variants of our PHP image. Combine it with our [Nginx image](/chainguard/chainguard-images/reference/nginx) and an optional database for a traditional LEMP setup.

> The overall migration process is essentially the same as described in the previous section, with the difference that you won't set up application entry points, since these images run as services. Your Dockerfile may require additional steps to set up front end dependencies, initialize databases, and perform any additional tasks needed for the application to run through a web server.

Reviewer: ```suggestion
The overall migration process is essentially the same as described in the previous section, with the difference that you won't set up application entry points, since these images run as services. Your Dockerfile may require additional steps to set up front-end dependencies, initialize databases, and perform any additional tasks needed for the application to run through a web server.
```

based mostly on [this](https://medium.com/@ist.stevkovski/is-it-front-end-or-front-end-or-frontend-3ae717cae4aa) which is not a reliable grammar reference but it looks right to me

      MARIADB_DATABASE: php-test
    ports:
      - 3306:3306
    networks:
      - wolfi

networks:
  wolfi:
    driver: bridge

```

Notice that the environment creates a few shared volumes to share the application files and also the `nginx.conf` file with the specific web server configuration. You'll need that to tell Nginx to redirect `.php` requests to the `php-fpm` service.

An example `nginx.conf` file to handle PHP requests with default settings:

```yaml
pid /var/run/nginx.pid;

events {
  worker_connections  1024;
}

http {
    server {
        listen 8080;
        index index.php index.html;
        root /app/laravel-demo/public;
        location ~ \.php$ {
            try_files $uri =404;
            fastcgi_split_path_info ^(.+\.php)(/.+)$;
            fastcgi_pass app:9000;
            fastcgi_index index.php;
            include fastcgi_params;
            fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
            fastcgi_param PATH_INFO $fastcgi_path_info;
        }
        location / {
            try_files $uri $uri/ /index.php?$query_string;
            gzip_static on;
        }
    }
}
```

## Additional Resources

> Our [PHP image documentation](/chainguard/chainguard-images/reference/php/) covers details about all PHP image variants, including the list of available tags for both development and production images. For another example of LEMP setup using MariaDB, check our [Getting Started with the MariaDB Chainguard Image](https://edu.chainguard.dev/chainguard/chainguard-images/getting-started/mariadb/).

Reviewer: ```suggestion
Our [PHP image documentation](/chainguard/chainguard-images/reference/php/) covers details about all PHP image variants, including the list of available tags for both development and production images. For another example of a LEMP setup using MariaDB, check our guide on [Getting Started with the MariaDB Chainguard Image](https://edu.chainguard.dev/chainguard/chainguard-images/getting-started/mariadb/).
```

> The [Debugging Distroless](/chainguard/chainguard-images/debugging-distroless-images/) guide contains important information for debugging issues with distroless images. Check also the [Verifying Images](/chainguard/chainguard-images/verifying-chainguard-images-and-metadata-signatures-with-cosign/) resource for details around provenance, SBOMs, and image signatures.

Reviewer: ```suggestion
The [Debugging Distroless](/chainguard/chainguard-images/debugging-distroless-images/) guide contains important information for debugging issues with distroless images. You can also refer to the [Verifying Images](/chainguard/chainguard-images/verifying-chainguard-images-and-metadata-signatures-with-cosign/) resource for details around provenance, SBOMs, and image signatures.
```

"check also" is kind of a funny way to start a sentence
